### PR TITLE
Forms

### PR DESCRIPTION
**Pull request by MarinaStepanchuk**
Task: https://github.com/rolling-scopes-school/tasks/tree/master/react/modules/module02
Screenshot: 
Screenshot_16

Screenshot_19

Screenshot_18

Deploy: [movie-search-rs.netlify.app](https://movie-search-rs.netlify.app/)
Done 26.03.2023 / deadline 27.03.2023
Score: 15 / 15

### CODE REVIEW

File: src/components/Form/Form.test.tsx [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+  it('should render correct', () => {
+    const textBoxes = screen.getAllByRole('textbox');
+    expect(textBoxes[0].getAttribute('id')).toBe(InputName.name);
+    expect(textBoxes[1].getAttribute('id')).toBe(InputName.movie);
+    expect(textBoxes[2].getAttribute('id')).toBe(InputName.review);
+    expect(screen.getByRole('combobox').getAttribute('id')).toBe(InputName.country);
+    expect(screen.getByTestId('dataInput').getAttribute('id')).toBe(InputName.photo);
+    expect(screen.getAllByRole('radio').length).toBe(2);
+    expect(screen.getByRole('checkbox').getAttribute('checked')).toBeFalsy();
```
Comment (Ksarelto) [inline src/components/Form/Form.test.tsx:26]: as I said 1 test => 1-2 expect not 7))

File: src/components/Form/Form.test.tsx [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+      photo: '',
+      date: '',
+      movie: '',
+      review: 'Good movie',
+      recommended: true,
+      unrecommended: false,
+      personal: false,
+    };
+
+    expect(onSubmit).toBeCalledWith(expectedCard);
```
Comment (Ksarelto) [inline src/components/Form/Form.test.tsx:60]: if u have one expect in the begining and one more in the end of the test that means there are 2 tests

File: src/pages/Error/Error.test.tsx [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import Error from './Error';
+import { render, screen } from '@testing-library/react';
+import React from 'react';
```
Comment (Ksarelto) [inline src/pages/Error/Error.test.tsx:3]: this import is not needed

File: src/main.tsx [hunk 4]
```diff
@@ -0,0 +1,13 @@
+import React from 'react';
+import ReactDOM from 'react-dom/client';
+import App from '../src/App/App';
+import './index.css';
```
Comment (Ksarelto) [inline src/main.tsx:4]: imports of css are the last in list of imports

File: src/constants/common.constants.tsx [hunk 5]
```diff
@@ -0,0 +1,335 @@
...
+  error: '404 Page not found',
+  critiques: 'Critiques',
+};
+
+const LocalStorageKeys = {
+  search: 'search',
+};
+
+const Countries = [
+  'Afghanistan',
```
Comment (Ksarelto) [inline src/constants/common.constants.tsx:75]: maybe better to use some npm package for countries list?

File: src/components/SearchBar/SearchBar.tsx [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+import classes from './SearchBar.module.scss';
+
+interface ISearchBarProps {
+  text: string;
+  onTextChange: (value: string) => void;
+}
+
+class SearchBar extends React.Component<ISearchBarProps> {
+  private handleTextChange: (e: React.ChangeEvent<HTMLInputElement>) => void = (
+    e: React.ChangeEvent<HTMLInputElement>
```
Comment (Ksarelto) [inline src/components/SearchBar/SearchBar.tsx:12]: u can use or first type or the second type defenition

File: src/components/SearchBar/SearchBar.tsx [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+  };
+
+  public render(): JSX.Element {
+    return (
+      <div className={classes.searchBar}>
+        <input
+          className={classes.search}
+          type="text"
+          value={this.props.text}
+          onChange={this.handleTextChange.bind(this)}
```
Comment (Ksarelto) [inline src/components/SearchBar/SearchBar.tsx:24]: try to not use binding

File: src/components/SearchBar/SearchBar.test.tsx [hunk 8]
```diff
@@ -0,0 +1,16 @@
+import SearchBar from './SearchBar';
+import { render, screen } from '@testing-library/react';
+import { vitest } from 'vitest';
+import userEvent from '@testing-library/user-event';
+
+describe('SearchBar', () => {
+  const onTextChange = vitest.fn();
+
+  it('call function test', async () => {
```
Comment (Ksarelto) [inline src/components/SearchBar/SearchBar.test.tsx:9]: 'should call...'

File: src/components/ItemMovie/ItemMovie.tsx [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+class ItemMovie extends React.Component<IItemMoviesProps> {
+  public render(): JSX.Element | undefined {
+    const { name, poster, year, countries, rating } = this.props.movie;
+    const text = this.props.text;
+
+    if (
+      !name.includes(text) &&
+      !year.toString().includes(text) &&
+      !rating.toString().includes(text) &&
+      countries.filter((country) => country.name.includes(text)).length === 0
```
Comment (Ksarelto) [inline src/components/ItemMovie/ItemMovie.tsx:21]: there are too many checks, better to create a separate funtion or method that will check all this conditions

File: src/components/FormItemWrapper/FormItemWrapper.tsx [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+  name: string;
+  children: ReactNode;
+  errorMessage?: string;
+  class?: string;
+}
+
+class FormItemWrapper extends React.Component<IFormItemProps> {
+  render(): JSX.Element | undefined {
+    const { label, name, children, errorMessage } = this.props;
+    return (
```
Comment (Ksarelto) [inline src/components/FormItemWrapper/FormItemWrapper.tsx:15]: empty line before return

File: src/pages/Critiques/Critiques.tsx [hunk 11]
```diff
@@ -0,0 +1,131 @@
...
+
+    if (movie.length === 0) {
+      errors.movie = ErrorMessages.emptyLine;
+    }
+
+    if (new Date(date).getTime() > new Date().getTime()) {
+      errors.date = ErrorMessages.wrongDate;
+    }
+
+    if (date.length === 0) {
```
Comment (Ksarelto) [inline src/pages/Critiques/Critiques.tsx:89]: can u use here?:
```
if (!date.length) {}
```

**Review by Ksarelto — COMMENTED** [review]
some mistakes are presented in several places, pleace check all code)

File: .eslintrc.json [hunk 12]
```diff
@@ -0,0 +1,53 @@
...
+    },
+    "root": true,
+    "extends": [
+        "plugin:react/recommended",
+        "plugin:@typescript-eslint/recommended",
+        "prettier",
+        "plugin:react-hooks/recommended",
+        "plugin:prettier/recommended"
+    ],
+    "overrides": [
```
[the remark this answers is not in the document]
Comment (MarinaStepanchuk) [reply .eslintrc.json:16]: for beauty?)) removed)

File: .eslintrc.json [hunk 13]
```diff
@@ -0,0 +1,53 @@
...
+    "parser": "@typescript-eslint/parser",
+    "parserOptions": {
+        "ecmaFeatures": {
+           "jsx": true
+        },
+        "ecmaVersion": "latest",
+        "sourceType": "module"
+    },
+    "plugins": [
+        "@typescript-eslint", "react", "prettier", "react-hooks"
```
[the remark this answers is not in the document]
Comment (MarinaStepanchuk) [reply .eslintrc.json:27]: ok, corrected

File: .eslintrc.json [hunk 14]
```diff
@@ -0,0 +1,53 @@
...
+        "@typescript-eslint/interface-name-prefix": "off",
+        "@typescript-eslint/ban-ts-comment": "error",
+        "@typescript-eslint/no-non-null-assertion": "off",
+        "@typescript-eslint/explicit-module-boundary-types": "off",
+        "@typescript-eslint/no-empty-function": "off",
+        "@typescript-eslint/no-explicit-any": "error",
+        "@typescript-eslint/no-var-reqiures": "off",
+        "react/jsx-uses-react": "off",
+        "comma-dangle": ["error", "only-multiline"],
+        "react/react-in-jsx-scope": "off"
```
[the remark this answers is not in the document]
Comment (MarinaStepanchuk) [reply .eslintrc.json:45]: corrected

File: src/components/Form/Form.module.scss [hunk 15]
```diff
@@ -0,0 +1,114 @@
...
+
+  .formBlock {
+    display: flex;
+    flex-direction: column;
+    gap: 10px;
+    align-items: flex-start;
+  }
+
+  input:not([type="file"]),
+  textarea,
```
[the remark this answers is not in the document]
Comment (MarinaStepanchuk) [reply src/components/Form/Form.module.scss:65]: added class

File: src/pages/Error/Error.test.tsx (hunk 3 above)
Comment (MarinaStepanchuk) [reply src/pages/Error/Error.test.tsx:3]: looked it over when I cleaned it, thanks. deleted

File: src/main.tsx (hunk 4 above)
Comment (MarinaStepanchuk) [reply src/main.tsx:4]: corrected

File: src/constants/common.constants.tsx (hunk 5 above)
Comment (MarinaStepanchuk) [reply src/constants/common.constants.tsx:75]: accidentally run into a not very smart inspector who will decide that I have connected something forbidden. Must use next time)

File: src/components/SearchBar/SearchBar.tsx (hunk 7 above)
Comment (MarinaStepanchuk) [reply src/components/SearchBar/SearchBar.tsx:24]: forgot to remove)

File: src/components/SearchBar/SearchBar.tsx (hunk 6 above)
Comment (MarinaStepanchuk) [reply src/components/SearchBar/SearchBar.tsx:12]: really)

File: src/components/SearchBar/SearchBar.test.tsx (hunk 8 above)
Comment (MarinaStepanchuk) [reply src/components/SearchBar/SearchBar.test.tsx:9]: corrected

File: src/components/ItemMovie/ItemMovie.tsx [hunk 16]
```diff
@@ -0,0 +1,50 @@
+import React from 'react';
```
[the remark this answers is not in the document]
Comment (MarinaStepanchuk) [reply src/components/ItemMovie/ItemMovie.tsx:1]: Yes, I'll take it into account next time, thanks)

File: src/components/FormItemWrapper/FormItemWrapper.tsx (hunk 10 above)
Comment (MarinaStepanchuk) [reply src/components/FormItemWrapper/FormItemWrapper.tsx:15]: added

File: src/pages/Critiques/Critiques.tsx (hunk 11 above)
Comment (MarinaStepanchuk) [reply src/pages/Critiques/Critiques.tsx:89]: Yes, sure)

File: src/components/Form/Form.tsx [hunk 17]
```diff
@@ -0,0 +1,199 @@
...
+    if (this.props.shouldClear) {
+      this.form.current?.reset();
+    }
+  }
+
+  public render(): JSX.Element | undefined {
+    return (
+      <form
+        className={classes.critiqueForm}
+        onSubmit={(event) => event.preventDefault()}
```
[the remark this answers is not in the document]
Comment (MarinaStepanchuk) [reply src/components/Form/Form.tsx:89]: moved to a separate function)

File: src/components/Header/Header.tsx [hunk 18]
```diff
@@ -0,0 +1,73 @@
+import React from 'react';
+import { Pages, Paths } from '../../constants/common.constants';
+import { NavLink } from 'react-router-dom';
+import classes from './Header.module.scss';
+
+const getRouteTitle = () => {
+  const url = location.pathname;
+  return Object.values(Paths).find((item) => item.path === url)?.title || '';
+};
```
[the remark this answers is not in the document]
Comment (MarinaStepanchuk) [reply src/components/Header/Header.tsx:9]: so you can mock my folder again?))) transferred)

File: src/components/Form/Form.test.tsx (hunk 2 above)
Comment (MarinaStepanchuk) [reply src/components/Form/Form.test.tsx:60]: divided tests

File: src/components/ItemMovie/ItemMovie.tsx (hunk 9 above)
Comment (MarinaStepanchuk) [reply src/components/ItemMovie/ItemMovie.tsx:21]: put in a separate method

**Pull request merged**